jackrabbit-jcr2spi/src/test/java/org/apache/jackrabbit/jcr2spi/hierarchy/ChildNodeEntriesImplTest.java

```diff
@@ -0,0 +1,39 @@
+package org.apache.jackrabbit.jcr2spi.hierarchy;
+
+import static org.junit.Assert.fail;
+import static org.mockito.Mockito.mock;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.ConcurrentModificationException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+import org.junit.Test;
+
+public class ChildNodeEntriesImplTest {
+
+    private final ExecutorService executorService = Executors.newFixedThreadPool(10);
+
+    @Test(expected = ConcurrentModificationException.class)
+    public void concurrentModification() throws Throwable {
+        ChildNodeEntriesImpl entries = new ChildNodeEntriesImpl(null, null, null);
+        Collection<Future<?>> futures = new ArrayList<>();
```

Comment:
@rkovarik the test failing looks like a timing issue. If you pre-populate the child nodes entries it has a greater chance to actually run into a ConcurrentModificationException. 
I.e. adding 100 child nodes did the trick for me. 

        for (int k = 0; k < 100; k++) {
            entries.add(mock(NodeEntry.class));
        }

Reply:
Right, thx @mduerig. I get only up to 1000 successful runs so I will remove the test, the code looks fine.

Reply:
I'd still like to see a test (that we can make fail by removing the modified checks in the old code) before proceeding.

Reply:
OK, pushed it back, ~one failure in 10000 runs. 
<img width="1051" alt="Screen Shot 2022-10-05 at 9 52 49" src="https://user-images.githubusercontent.com/14125009/194030046-46b36c25-f7ed-41f2-93d6-eae454a1d3ba.png">

Reply:
The test is non deterministic so I'd expect it to fail from time to time. I don't see a way to implement a deterministic test here unless we make `LinkedEntries` package private and unit test that class directly.

Reply:
I already moved it into a separate class (see https://issues.apache.org/jira/browse/JCR-4854)